### PR TITLE
Keep snapped coordinates when using vertex tool on a curved geometries

### PR DESCRIPTION
**Pull request by alexbruy**
## Description

When using vertex tool to modify geometry and snap to a curve, the rubber band shows a correct snap but after "commiting" changes updated point jumps to a location on a segmentized curve.

This happens because `QgsPointLocator::Match::interpolatedPoint()` performs interpolation on a segment, even if curved geometry is used. Proposed fix keeps snapped X and Y coordinates while Z coordinate is interpolated from a segment.

Refs #53197.

Funded by NetPMD

### CODE REVIEW

**Review by troopa81 — COMMENTED** [review]
Wouldn't be better to do it in QgsGeometryUtils::closestPoint ? And I think Z/M should be computed according to the distance with the snapped point, not the false segment interpolated point, no? 

Do you mind adding a test?

File: src/core/qgspointlocator.h [hunk 1]
```diff
@@ -287,11 +287,20 @@ class CORE_EXPORT QgsPointLocator : public QObject
 
             if ( !( geom.isNull() || geom.isEmpty() ) )
             {
-              const QgsLineString line( geom.vertexAt( mVertexIndex ), geom.vertexAt( mVertexIndex + 1 ) );
-              point = QgsGeometryUtils::closestPoint( line, QgsPoint( snappedPoint ) );
+              // when snapping to a curve we need to use real geometry in order to have correct location
+              // of the snapped point, see https://github.com/qgis/QGIS/issues/53197.
+              // In other cases it is ok to use only a segment to speedup calculations.
+              if ( QgsWkbTypes::isCurvedType( mLayer->wkbType() ) )
```
Comment (uclaros) [inline src/core/qgspointlocator.h:293]: Could we not use `geom.wkbType()` here so the optimized route is used for non-curved geometries within curved layers?

Comment (alexbruy) [reply src/core/qgspointlocator.h]: Changed to use `geom.wkbType()`

**Review by uclaros — APPROVED** [review]

**Review by troopa81 — APPROVED** [review]

**Comment by lbartoletti** [comment]
@alexbruy @uclaros @troopa81 Do you think there is a bug in `QgsPointLocator::Match::interpolatedPoint()`? Snapping a point on a curve, with the vertex tool or simply adding a point, result in an offset snapping point:

Vertex tool : 


https://github.com/qgis/QGIS/assets/7521540/9fe724cb-d1d8-4f21-88ba-7ddff844fdd7



Add a point :


https://github.com/qgis/QGIS/assets/7521540/08d5efdd-a917-40c8-a724-3c8c1adec5db

Obviously, with a straight version of this curve, the snapping is ok.

**Comment by uclaros** [comment]
What's the offset distance in units? (Scale is not visible in your videos.)
If it's very small, it could as well be the curved rendering that is actually off.

**Comment by lbartoletti** [comment]
> What's the offset distance in units? (Scale is not visible in your videos.) If it's very small, it could as well be the curved rendering that is actually off.

More than 10 cm on EPSG:3946 
The geom: 'SRID=3946;CompoundCurve (CircularString (1973422.15771776507608593 5198767.19862847682088614, 1976139.65683170617558062 5200645.02104990836232901, 1979620.49839338357560337 5199698.47641471493989229))' 

image

**Comment by uclaros** [comment]
I think this is a rendering issue, as arcs are rendered segmentized by QT. When dealing with big arcs, the offset of the segments can be significant.
Check how your arc is rendered against the 3 points that define it:

image

At that point the offset is around 1.5mm, meaning it's closer to a segment's vertex than in your screenshot.

**Pull request merged**